crates/ra_hir_def/src/expr.rs

```diff
@@ -376,35 +376,14 @@ pub enum Pat {
     Wild,
     Tuple(Vec<PatId>),
     Or(Vec<PatId>),
-    Record {
-        path: Option<Path>,
-        args: Vec<RecordFieldPat>,
-        // FIXME: 'ellipsis' option
-    },
-    Range {
-        start: ExprId,
-        end: ExprId,
-    },
-    Slice {
-        prefix: Vec<PatId>,
-        slice: Option<PatId>,
-        suffix: Vec<PatId>,
-    },
+    Record { path: Option<Path>, args: Vec<RecordFieldPat>, ellipsis: Option<PatId> },
```

Comment:
I wonder if we really should give ellipsis it's own pat_id.... Since recently, syntactically it is indeed a completely separate pattern kind. Semantically, however, it only fits in records and tuples, and can be represented as

```rust
Record { path: Option<Path>, args: Vec<RecordFieldPat>, ellipsis: bool },
TupleStruct { path: Option<Path>, args: Vec<PatId>, ellipsis: Option<usize> },
```

I guess, in theory we can even elaborate an `...` to `foo: _, bar: _, baz: _` *after* name resolution, although we don't have that IR level, and we can't really conveniently elaborate-away a small bit of expressions grammar.

Reply:
I agree, the ellipsis is not a pattern, that it is a pattern in the AST is just an artifact of parsing.

Reply:
> `Record { path: Option<Path>, args: Vec<RecordFieldPat>, ellipsis: bool }`

I actually implemented it this way (with a bool) at first, and then later switched to the `PatId` since I thought we may want to be able to link this back to the AST in case of creating a diagnostic.

```rust
struct Foo {
    a: (),
}

fn bar(foo: Foo) {
    let Foo { .., a } = foo; 
    // I think we may eventually want to show a diagnostic on
    // the ellipsis pattern here?
    // rustc says "`..` must be at the end and cannot have a trailing comma"
}
```

If we store `ellipsis` as a bool, is there some other way to create this diagnostic on the ellipsis?

Reply:
Yes, we can still find the ellipsis in the AST if we need to put a diagnostic on it.

Reply:
>I actually implemented it this way (with a bool) at first, and then later switched to the PatId since I thought we may want to be able to link this back to the AST in case of creating a diagnostic.

Yup, I've also thought along those lines (hence "I wonder" and not "we should"). I've thought a bit more about it, and now I am pretty sure that we should go with the `bool`. More generally, I think we should push really hard on making `Expr` as simple as desugared as possible (and, for example, remove the `Expr::For`), and push all source <-> lowered expression info into a side-table. Admittedly, such sidetable does not exist yet, but the harder we make it to get the source, the more pressure we'll have to actually add something for that purpose :)

Reply:
Sounds good to me. Thanks for providing the additional insight here!

I've changed this to a bool.